### PR TITLE
markdown: add document listing TPM2 PCR measurements we make comprehensively

### PR DESCRIPTION
**Pull request by poettering**
This is useful to write TPM event log decoders.

### CODE REVIEW

**Comment by poettering** [comment]
(in case this wasn't clear, I wrote this down while working on #28891)

File: docs/TPM2_PCR_MEASUREMENTS.md [hunk 1]
```diff
@@ -0,0 +1,137 @@
+---
+title: Recognizing TPM2 PCR Measurements Made by systemd
```
Comment (keszybz) [inline docs/TPM2_PCR_MEASUREMENTS.md]: The text is about "TPM2 PCR measurements made by systemd" — the first para makes this clear. So drop "Recognizing" here.

File: docs/TPM2_PCR_MEASUREMENTS.md [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+category: Booting
+layout: default
+SPDX-License-Identifier: LGPL-2.1-or-later
+---
+
+# Recognizing TPM2 PCR Measurements Made by systemd
```
Comment (keszybz) [inline docs/TPM2_PCR_MEASUREMENTS.md]: Also here.

File: docs/TPM2_PCR_MEASUREMENTS.md [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+Various systemd components issue TPM2 PCR measurements during the boot process,
+both in UEFI mode and from userspace. The following lists all measurements
+done, and describes (in case done before `ExitBootServices()`) how they appear
+in the TPM2 Event Log, maintained by the PC firmware.
+
+systemd will measure to PCRs 11, 12, 13, 15.
+
+Note that the userspace measurements listed below are only done on systems
+booted with `systemd-stub`.
```
Comment (keszybz) [inline docs/TPM2_PCR_MEASUREMENTS.md]: This third paragraph should be folded into the first paragraph to form a coherent whole. Also, it's quite unclear how the measurements are done both in UEFI mode an in userspace if they're only done by `systemd-stub`.

File: docs/TPM2_PCR_MEASUREMENTS.md [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+---
+
+# Recognizing TPM2 PCR Measurements Made by systemd
+
+Various systemd components issue TPM2 PCR measurements during the boot process,
+both in UEFI mode and from userspace. The following lists all measurements
+done, and describes (in case done before `ExitBootServices()`) how they appear
+in the TPM2 Event Log, maintained by the PC firmware.
+
+systemd will measure to PCRs 11, 12, 13, 15.
```
Comment (keszybz) [inline docs/TPM2_PCR_MEASUREMENTS.md]: We now have official names for the PCRs, so I think this should say
"foobar (11), barbar (12), …."

File: docs/TPM2_PCR_MEASUREMENTS.md [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+
+If the kernel command line was specified explicitly (by the user or in a Boot
+Loader Specification Type #1 file), the kernel command line passed to the
+invoked kernel is measured before it is executed.
+
+*Description* in the event log record is the literal kernel command line in
+UTF-16.
+
+*Measured hash* covers the literal kernel command line in UTF-16 (without any
+trailing NUL byte).
```
Comment (keszybz) [inline docs/TPM2_PCR_MEASUREMENTS.md]: "bytes". Also below.

File: docs/TPM2_PCR_MEASUREMENTS.md [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+in the TPM2 Event Log, maintained by the PC firmware.
+
+systemd will measure to PCRs 11, 12, 13, 15.
+
+Note that the userspace measurements listed below are only done on systems
+booted with `systemd-stub`.
+
+## PCR Measurements Made by `systemd-boot` (UEFI)
+
+### PCR 12, EV_IPL, "Kernel Command Line"
```
Comment (keszybz) [inline docs/TPM2_PCR_MEASUREMENTS.md]: `EV_IPL` (add markup). Also below.

File: docs/TPM2_PCR_MEASUREMENTS.md [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+
+If the kernel command line was specified explicitly (by the user or in a Boot
+Loader Specification Type #1 file), the kernel command line passed to the
+invoked kernel is measured before it is executed.
```
Comment (keszybz) [inline docs/TPM2_PCR_MEASUREMENTS.md]: Maybe add a parenthesized remark here that for Type 2 the command line is measured as part of the UKI.

File: docs/TPM2_PCR_MEASUREMENTS.md [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+UTF-16.
+
+*Measured hash* covers the literal kernel command line in UTF-16 (without any
+trailing NUL byte).
+
+## PCR Measurements Made by `systemd-stub` (UEFI)
+
+### PCR 11, EV_IPL "PE Section Name"
+
+Happens once for each UKI-defined PE section of the UKI, in the canonical UKI
```
Comment (keszybz) [inline docs/TPM2_PCR_MEASUREMENTS.md:59]: "A measurement is made for each PE section of the UKI that is described by the UKI specification (link), in the canonical order described in the specification."

File: docs/TPM2_PCR_MEASUREMENTS.md [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+PE section order, as per the UKI specification. For each record a pair of
+records are written, first one that covers the PE Section Name, and then one
```
Comment (keszybz) [inline docs/TPM2_PCR_MEASUREMENTS.md]: "pair" is singular, so "a pair of records is written".

"is written: first one …, and the second one …".

"section name", "section data" (lowercase).

File: docs/TPM2_PCR_MEASUREMENTS.md [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+that covers the "PE Section Data", so that both types of records appear
+interleaved in the event log.
+
+*Description* in the event log record is the PE section name in UTF-16.
+
+*Measured hash* covers the PE section name in UTF-8 (including a trailing NUL byte!).
```
Comment (keszybz) [inline docs/TPM2_PCR_MEASUREMENTS.md]: ~~"section name or section data"? But the trailing NUL byte does not apply to section data?~~

Oh, so this section describes just the first of the pair of measurements. This is very confusing.

File: docs/TPM2_PCR_MEASUREMENTS.md [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+
+## PCR Measurements Made by `systemd-pcrextend` (Userspace)
+
+### PCR 11, "Boot Phases"
+
+The `systemd-pcrphase.service`, `systemd-pcrphase-initrd.service`,
+`systemd-pcrphase-sysinit.service` services will measure the boot phase reached
+during various times of the boot process. Specifically, the strings
+"enter-initrd", "leave-initrd", "sysinit", "ready", "shutdown", "final" are
+measured, in this order.
```
Comment (keszybz) [inline docs/TPM2_PCR_MEASUREMENTS.md]: Maybe mention that those are normal units so the distribution or local adminstrator may modify the list the of measured phases?

File: docs/TPM2_PCR_MEASUREMENTS.md [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+
+*Measured hash* covers the listed phase strings (in UTF-8).
```
Comment (keszybz) [inline docs/TPM2_PCR_MEASUREMENTS.md]: "covers the phase string (in UTF-8)". What about the NUL byte?

File: docs/TPM2_PCR_MEASUREMENTS.md [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+
+### PCR 15, "Machine ID"
+
+The `systemd-pcrmachine.service` service will measure the machine ID (as read
+from `/etc/machine-id`) during boot.
+
+*Measured hash* covers the string "machine-id:" suffixed by the machine ID
+formatted in hexadecimal lowercase characters.
```
Comment (keszybz) [inline docs/TPM2_PCR_MEASUREMENTS.md]: What about the NUL byte?

File: docs/TPM2_PCR_MEASUREMENTS.md [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+
+### PCR 15, "File System"
+
+The `systemd-pcrfs-root.service` and `systemd-pcrfs@.service` services will
+measure a string identifying a specific file system, typically covering the
+root file system and `/var/` (if it is its own file system).
+
+*Measured hash* covers the string "file-system:" suffixed by a series of six
+colon-separated strings, identifying the file system type, UUID, label as well
+as the GPT partition entry UUID, entry type UUID and entry label.
```
Comment (keszybz) [inline docs/TPM2_PCR_MEASUREMENTS.md]: What about the NUL byte?

File: docs/TPM2_PCR_MEASUREMENTS.md [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+## PCR Measurements Made by `systemd-cryptsetup` (Userspace)
+
+### PCR 15, "Volume Key"
+
+The `systemd-cryptsetup@.service` service will measure a key derived from the
+LUKS volume key of a specific encrypted volume, typically covering the backing
+encryption device of the root file system and `/var/` (if it is its own file
+system).
+
+*Measured hash* covers the result of the HMAC(V,S) calculation where V := LUKS
```
Comment (keszybz) [inline docs/TPM2_PCR_MEASUREMENTS.md]: Change ":=" to is?

**Review by keszybz — COMMENTED** [review]
Looks reasonable, but could use a little polish for readibility and consistency.

**Comment by poettering** [comment]
Addressed all issues raised by @keszybz

**Pull request merged**